### PR TITLE
DS-3909: Adding withdraw, reinstate, and isDiscoverable support to item repository.

### PR DESCRIPTION
**Pull request by mspalti**
@abollini  I have a few questions about this work. I think I understand that the task generally. I'm not sure that I've handled exceptions correctly.  And now looking at the documentation at:
 
https://github.com/DSpace/Rest7Contract/blob/master/patch.md
https://github.com/DSpace/Rest7Contract/blob/master/workspaceitem-data-metadata.md

Which again makes me think that we need additional error handling if it's not already provided by default (e.g.: 405 Method Not Allowed)

For usage, see https://github.com/DSpace/Rest7Contract/pull/25/files

### CODE REVIEW

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (tdonohue) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: @mspalti : It looks like this PR accidentally performs a lot of code realignment (of existing code).  Is there any chance you can undo the code realignment?  It doesn't seem necessary for the PR, and it distracts from the actual code changes.  My guess is this might have been done automatically by your IDE or something, as the old realignment was perfectly OK by our code style.

Overall, this isn't a big deal, but it'd be a way to make this PR smaller overall.

Comment (mspalti) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: Yes, it was IDE code alignment.  Will fix.

**Comment by mspalti** [comment]
OK, now that a patch method is working I need to ask a few questions...

I see that the submit workflow uses a factory class and various implementations for patch operations.  I've kept things simple so far, but I'm wondering if a different approach will be helpful here?  

I'm also now looking at https://github.com/mspalti/Rest7Contract/blob/master/patch.md and seeing requirements I hadn't considered.  That makes me wonder if the simple, local patch method as I've defined it is a good idea.  It might be. I'm just thinking out loud here.

**Comment by mspalti** [comment]
This answers my question!  I just saw this on the Dspace 7 sprint page. This is how I thought I should approach the problem but wasn't sure since I didn't see it mentioned in the instructions for DS-3909. 

... implement the patch method in the Item repository using a support class so to be reusable for each DSpaceObject (Collection, Community, EPerson, ...)

**Comment by mspalti** [comment]
Changes

- Added base class to handle all resource patch requests (`AbstractResourcePatch`). This is a template that includes a patch method and default methods for each operation (add, remove, replace, move). The default methods throw an error (`PatchUnprocessableEntityException`).
- Added a patch implementation class (`ItemPatch`) and annotated it to be a Spring component.  Autowired the component into `ItemRestRepository`.
- `ItemPatch` implements concrete methods for `withdraw`, `reinstate`, and `discoverable` operation paths.
- It is assumed that patch operations for other resource repositories (e.g. `EPersonRestRepository`) will implement classes that extend the base template.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java]: This is probably not needed. I have discovered that the HAL browser doesn't send the Authorization header for method other than GET see https://jira.spring.io/browse/DATAREST-1077
you should try to test using postman or command line tools like curl

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: a replace that send the current value should be just a no-op resulting in OK

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: better create separate test methods for the different scenarios
- withdrawn
- reinstatiate
- private
- unprivate (discoverable)

**Review by abollini — CHANGES_REQUESTED** [review]

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
Comment (mspalti) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java]: You are right. The patch is successful when the Authorization header is included.  Removing the temporary fix.

**Comment by mspalti** [comment]
Changes:

- Separate integration tests for withdraw, reinstate, discover and private patch requests.
- Builder classes for collection and item mocks have new methods for setting ADD and REMOVE permissions on collections and WRITE permissions on items (without these permissions tests were failing).

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/ItemPatch.java
Comment (mspalti) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/ItemPatch.java]: This is not implemented because I'm not sure what to test for. What does it mean for something not to be archived?  Should I check the "isArchived" flag in the item or is there something else that I should look for?

**Comment by mspalti** [comment]
I think I'm ready to stop working on this and wait for review.

In addition to my question about how to test for whether or not in item is "archived" I'm also curious about the idea of using a template for the helper class rather than a factory (as was done with workspace items).  I'm happy to refactor to use a factory if that will scale better when more features are added to the api. But as of right now the template seems like a good solution to me because this PR doesn't have the added complexity of configurable workflows.

**Comment by tdonohue** [comment]
Hi @mspalti, thanks for your continued work here!  Just as an FYI, reviews might be a little more "spread out" in the coming weeks as both @abollini and I are heads down in major preparations for [OR2018](http://www.or2018.net/) (June 4-7).  After the conference we'll be much more available, but until then, we'll have to see where we can squeeze in quick reviews -- so, bear with us if we aren't able to respond immediately.

**Comment by mspalti** [comment]
Understood.  Hope you both enjoy the conference!

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java]: This method should be responsible to retrieve the item so to centralize the management of a 404 error and reduce code duplication.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java]: This class should use generics to deal with the RestModel instead than apiCategory, model and ID
i.e. `AbstractResourcePatch<R extends RestModel>` and use `R restModel` instead than the three arguments in the subsequent methods

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java]: a default implementation can be provided. According to the spec a replace operation is equivalent to a remove followed by an add operation

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: is this really needed? the getClient(token) already do that

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: it is not required but it could be easier to instantiate a Patch object and serialize it using a Jackson ObjectMapper.
The code will be more readable and consistent as I guess it will be hard to write more complex test by hand maybe involving multiple operations etc.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/AbstractDSpaceObjectBuilder.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/AbstractDSpaceObjectBuilder.java]: this is not necessary as the test can be done using an administrator user. Nevertheless, it is a good idea to introduce support for add/remove permissions but it will require more investigation as we need to be sure that these extra permissions (resourcepolicies) are also cleanup when the test end. It will be probably easier to delegate to a ResourcePolicy builder for that.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: instead than introduce support for write/remove permission please use an administrator account to run the test (see also comments below)

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: there is also an admin attribute in the abstract test class other than eperson that can be used to run the test as an administrator (or a makeAdministrator method - in previous commits).
Please note that there was a bug in the implementation that lead to issue when test for unauthorized access after have used the admin user that has been solved in #2073

**Review by abollini — CHANGES_REQUESTED** [review]
Hi Michael,
thank you for the PR and to continue to work on it.
I have added some changes request, let me know if you need help. I will be happy to make the changes for you as I really want to merge this PR soon. BTW some of the refactoring can be done also later if you prefer as a separate PR.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java
Comment (mspalti) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java]: Thanks, Andrea. I am just getting back to this now.  

I will work on your changes and let you know if I need help.  It's a chance for me to learn.  But if it looks too difficult I will let you know so you can jump in and get it finished fast!

**Comment by mspalti** [comment]
Hi Andrea,
I made the changes that you requested.  I'm available if needed to make more changes (and I think I understand this well enough now to help out).  Let me know.

You also mentioned that you want to merge the PR soon, so if it is easier for you, please feel free to make changes yourself. I'm ok with that!

Quick summary: Added generics to the abstract class and simplified method signatures.  Since you mentioned retrieving the item in the repository to centralize handling of 404, I changed the item repository patch method to throw a ResourceNotFoundException if it fails to locate the item. (Perhaps that wasn't necessary.) Made a few changes to the ItemPatch class to handle exceptions and get a copy of the DSpaceObject. Other changes to tests, etc.  

Some of the javadocs could be improved.

I did leave the code in AbstractDspaceObjectBuilder since you thought it might be of use, but I can remove that too if you think it best.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
Comment (mspalti) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java]: Andrea, this does not have the effect that I expected.  The controller's patch method doesn't specifically handle ResourceNotFoundException and instead returns 500.  

I guess to my way of thinking, if the item is not found one might return 422 rather than 404.  However, the controller can be modified to return 404 if that is the correct behavior.

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/AbstractDSpaceObjectBuilder.java
Comment (mspalti) [reply dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/AbstractDSpaceObjectBuilder.java]: should I remove this code now or leave it in place for further investigation? It sounds like we should remove it, but checking to be sure.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
Comment (mspalti) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java]: I modified to controller so that it throws the ResourceNotFoundException.

**Comment by mspalti** [comment]
Hi Andrea,
I completed work on the change requests.  Removed all of the unnecessary test code that I had added before I knew about the admin user.  Also, I added the 404 response the the repository patch controller.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java
Comment (mspalti) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java]: Andrea, would it be better to extend DSpaceObject?

**Comment by abollini** [comment]
The code looks good at this point, I will like to double check this code a last time after the merge of
https://github.com/DSpace/DSpace/pull/2025
as in the above PR I have fixed some issue with the use of the administrator user during the test
see for instance https://github.com/DSpace/DSpace/pull/2025/files#diff-6c13d89b51575219119d28ce0f97e225L208
and 
https://github.com/DSpace/DSpace/pull/2025/files#diff-39c111ad55a6fba18691ab2807a855e6L458
I will solve the current minor conflict and the new conflicts that the merge of PR#2025 will generate and merge the PR.

**Comment by abollini** [comment]
solved the conflict with the master, fixed authorization handling providing more test to verify.
Once travis will confirm that all is fine I will merge

**Review by abollini — APPROVED** [review]

**Pull request merged**
